React/testing.md

```diff
@@ -0,0 +1,234 @@
+
+In fact, in the past we had to install the testing library separately every time we created a new project with `create-react-app`, but now it comes bundled with `create-react-app`!  We will need to install watchman to run tests in guard mode (like Guard with Ruby).
+
+To run tests in React in guard-mode we'll need to first install watchman with `brew`.
+
+```bash
+$ brew install watchman
+```
+
+### Test Files
+
+We will use the following pattern for our test files:
+
+- Create a new folder `src/components/test`
+- Each component `src/components/X.js` gets a test file `src/components/test/X.test.js`
+- Jest should be able to automatically find and run any file ending in `.test.js` without us having to do any extra work
+
+Run `mkdir src/components/test` now to create this folder.
+
+## How to Run Tests
+
+Create-react-app comes with Jest testing built in and in fact provides an initial test.  To run the test type:  `npm run test` and you will get a screen like this:
+
+```bash
+No tests found related to files changed since last commit.
+Press `a` to run all tests, or run Jest with `--watchAll`.
+
+Watch Usage
+ › Press a to run all tests.
+ › Press p to filter by a filename regex pattern.
+ › Press t to filter by a test name regex pattern.
+ › Press q to quit watch mode.
+ › Press Enter to trigger a test run.
+ ```
+
+The test-runner will automatically run tests which have changed since the last save.  As you develop a React application, you can leave the tests running in a terminal tab and the tests will re-execute when you save files.
+
+  **Press `a` now** to manually re-run all the existing tests.
+
+ You should get the following:
+ ```bash
+ ... big error message
+
+Test Suites: 1 failed, 1 total
+Tests:       1 failed, 1 total
+Snapshots:   0 total
+Time:        3.34s
+Ran all test suites.
+
+Watch Usage: Press w to show more.
+```
+
+**OMG!!!** We're already failing a test!   That test file is `src/App.test.js`.
+
+Open that file:
+
+```javascript
+// src/App.test.js
+
+import React from 'react';
+import { render } from '@testing-library/react';
+import App from './App';
+
+test('renders learn react link', () => {
+  const { getByText } = render(<App />);
+  const linkElement = getByText(/learn react/i);
+  expect(linkElement).toBeInTheDocument();
+});
+```
+
+The test manually renders the `App` component and then tries to get some text with the words "learn react".
+
+Hmmm... we deleted that text...  
+
+**Question**  What do you notice?  What do you think is happening here?
+
+<details>
+  <summary>Walkthrough</summary>
+1.  The default test in `App.test.js` first renders an instance of the `App` component in a simulated DOM.  
```

Comment:
This is unfortunately not being formatted as a list, I think because it's in a details block. I remember running into this before but I don't remember if I found a solution... :/

Reply:
Fixed it by adding HTML, not ideal but it works.